ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -26,15 +90,69 @@ private[auth] final class OngoingAuthorizationObserver[A](
 
   override def request(i: Int): Unit = observer.request(i)
 
-  override def setMessageCompression(b: Boolean): Unit = observer.setMessageCompression(b)
+  override def setMessageCompression(b: Boolean): Unit =
+    self.synchronized(observer.setMessageCompression(b))
 
-  override def onNext(v: A): Unit =
-    authorized(claims) match {
-      case Right(_) => observer.onNext(v)
-      case Left(authorizationError) => observer.onError(throwOnFailure(authorizationError))
+  override def onNext(v: A): Unit = {
+    val now = nowF()
+    (for {
+      _ <- checkClaimsExpiry(now)
+      _ <- checkUserRightsRefreshTimeout(now)
+    } yield {
+      ()
+    }) match {
+      case Right(_) => self.synchronized(observer.onNext(v))
+      case Left(e) =>
+        cancelUserRightsCheckTask()
+        self.synchronized(observer.onError(e))
     }
+  }
 
-  override def onError(throwable: Throwable): Unit = observer.onError(throwable)
+  override def onError(throwable: Throwable): Unit = {
+    cancelUserRightsCheckTask()
+    self.synchronized(observer.onError(throwable))
+  }
+
+  override def onCompleted(): Unit = {
+    cancelUserRightsCheckTask()
+    self.synchronized(observer.onCompleted())
+  }
+
+  private def checkUserRightsRefreshTimeout(now: Instant): Either[StatusRuntimeException, Unit] = {
+    if (
+      originalClaims.resolvedFromUser &&
+      lastUserRightsCheckTime.isAfter(
+        now.plusSeconds(2 * userRightsCheckIntervalInSeconds.toLong)
```

Comment:
I think the `2 *` bit is a misunderstanding. We should check every `userRightsCheckIntervalInSeconds`. While doing so we guarantee that in the worst case we will be `2 * cacheExpiryAfterWriteInSeconds` behind the update

Reply:
Makes sense!

Reply:
Discussed offline as staying with `2 *`